circom/circom_test.go

```diff
@@ -0,0 +1,471 @@
+package circom
+
+import (
+	"fmt"
+	"github.com/consensys/gnark-crypto/ecc"
+	fr_bls12381 "github.com/consensys/gnark-crypto/ecc/bls12-377/fr"
+	bls12381 "github.com/consensys/gnark-crypto/ecc/bls12-381"
+	fr_bn254 "github.com/consensys/gnark-crypto/ecc/bn254/fr"
+	"github.com/consensys/gnark-crypto/ecc/secp256k1"
+	"github.com/consensys/gnark-crypto/ecc/secp256k1/fp"
+	fr_secp "github.com/consensys/gnark-crypto/ecc/secp256k1/fr"
+	groth16 "github.com/consensys/gnark/backend/groth16/bn254"
+	"github.com/consensys/gnark/backend/groth16/bn254/mpcsetup"
+	"github.com/consensys/gnark/constraint"
+	cs "github.com/consensys/gnark/constraint/bn254"
+	"github.com/consensys/gnark/frontend"
+	"github.com/consensys/gnark/frontend/cs/r1cs"
+	"github.com/consensys/gnark/std/algebra/emulated/sw_emulated"
+	zksha3 "github.com/consensys/gnark/std/hash/sha3"
+	"github.com/consensys/gnark/std/math/bits"
+	"github.com/consensys/gnark/std/math/emulated"
+	"github.com/consensys/gnark/std/math/uints"
+	"github.com/consensys/gnark/test"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/crypto/ecies"
+	"golang.org/x/crypto/sha3"
+	"math/big"
+	"math/rand"
+	"os"
+	"strconv"
+	"testing"
+	"time"
+)
+
+func doMPCSetUp(ccs constraint.ConstraintSystem) (pk groth16.ProvingKey, vk groth16.VerifyingKey, err error) {
+	const (
+		nContributionsPhase1 = 3
+		nContributionsPhase2 = 3
+		power                = 20 //19 //2^9 元素个数
+	)
+
+	initPhase1 := mpcsetup.InitPhase1(power)
```

Comment:
Also, do we really need to perform Phase 1 by ourselves in real life? There's existing completed MPC ceremonies that contain calculations for the large number of powers of tau. We used existing computation results when we integrated Groth16 for Neo.

Reply:
Other existing data can be used. However, compatibility issues may need to be considered, and some serialization methods may need to be implemented. This part of the entire MPC code will be split into several instructions and placed in the cli.At present, I use a loop to write it in one piece

Reply:
In addition, the amount of calculation in phase 1 is adjustable, and we do not need to use 2^27 like zcash does. The size I used here is 2^20r, and it only takes more than ten minutes to calculate.

Reply:
> This part of the entire MPC code will be split into several instructions and placed in the cli

We probably should consider reusing existing tools for the Phase 1, it will simplify things a bit.

> and we do not need to use 2^27 like zcash does

They do this to support circuits with large number of constraints. Are we going to use ceremony results only for this AES256 circuit? If so, then you're right, we don't need a lot of powers for it.

Reply:
I don’t think we need a huge amount because this part of the circuit is only used by our nodes themselves and does not need to be open to users, and our circuit is not as complex as zcash.

Reply:
Good, resolved.